File: fork.c
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <signal.h>

#define CNT 2000

void handler(int signo){
  static int count =0;
  count ++;
  printf("call number %d\n",count);
 
}

int main(void){
  signal (SIGCHLD,handler);
  int cnt = CNT;
  int *ids;
  ids = malloc  (sizeof(int)*CNT);
  while(--cnt>0)
  {
      int r = 2 + rand()%10;
    ids[cnt]=fork();
    if(ids[cnt] == 0)
    { 
//    printf("is child\n");
      sleep(r);
      //printf("rand=%d\n",r);
      return 0;
    }
   }
  while(1){ sleep(1); }
  return 0;
}

File: fork.php
<?php
declare(ticks=1);
$cnt=323;
$id=[];
pcntl_signal(SIGCHLD,'handlesignal');
pcntl_signal(SIGINT,'interrupt',true);
while(--$cnt>0){
  $pid = pcntl_fork();
  $id[$pid]='runing';
  if($pid===0){
    sleep(rand(4,8));
    exit(0);
  }
#  echo "child is $pid\n";
}

function handlesignal($signo){
 global $id;
    foreach($id as $pid=>$st){
      $stat='a';
      $ret = pcntl_waitpid($pid,$stat,WNOHANG);
      if($ret === -1){
        unset($id[$pid]);
      }
  }
  $prcnt = count($id);
  echo "RUNNING:::::: $prcnt\n";
  if(!$prcnt ) exit(0);  
  #pcntl_signal(SIGCHLD,'handlesignal',true);
}

function interrupt($sig){
 global $id;
 foreach($id as $pid=>$run){
  echo "sending SIGKILL to $pid ..... \n";
  posix_kill($pid,SIGINT);
 }
}

while(1) pcntl_wait($status);



##    RUNNING:::::: 34
##    RUNNING:::::: 33
##    RUNNING:::::: 32
##    RUNNING:::::: 31
##    RUNNING:::::: 16
##    RUNNING:::::: 5
##    RUNNING:::::: 4
##    RUNNING:::::: 4
##    

